cpp/src/io/json/json_tree.cu

```diff
@@ -380,6 +389,106 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
     stream);
   CUDF_EXPECTS(node_range_out_end - node_range_out_it == num_nodes, "node range count mismatch");
 
+  // Extract Struct, List range_end:
+  // 1. Extract Struct, List - begin & end separately, their token ids
+  // 2. push, pop to get levels
+  // 3. copy first child's parent token_id, also translate to node_id
+  // 4. propagate to siblings using levels, parent token id. (segmented scan)
+  // 5. scatter to node_range_end for only nested end tokens.
+  if (is_strict_nested_boundaries) {
+    // Whether the token is nested
+    auto const is_nested = [] __device__(PdaTokenT const token) -> bool {
+      switch (token) {
+        case token_t::StructBegin:
+        case token_t::StructEnd:
+        case token_t::ListBegin:
+        case token_t::ListEnd: return true;
+        default: return false;
+      };
+    };
+    auto const num_nested =
+      thrust::count_if(rmm::exec_policy(stream), tokens.begin(), tokens.end(), is_nested);
+    rmm::device_uvector<TreeDepthT> token_levels(num_nested, stream);
+    rmm::device_uvector<NodeIndexT> token_id(num_nested, stream);
+    rmm::device_uvector<NodeIndexT> parent_node_ids(num_nested, stream);
+    auto const push_pop_it = thrust::make_transform_iterator(
+      tokens.begin(),
+      cuda::proclaim_return_type<cudf::size_type>([] __device__(PdaTokenT const token) {
+        int const is_begin = token == token_t::StructBegin or token == token_t::ListBegin;
+        int const is_end   = token == token_t::StructEnd or token == token_t::ListEnd;
+        return is_begin - is_end;
+      }));
+    // copy_if only struct/list's token levels, token ids, tokens.
+    cudf::detail::copy_if_safe(push_pop_it,
+                               push_pop_it + num_tokens,
+                               tokens.begin(),
+                               token_levels.begin(),
+                               is_nested,
+                               stream);
+    cudf::detail::copy_if_safe(thrust::make_counting_iterator<NodeIndexT>(0),
+                               thrust::make_counting_iterator<NodeIndexT>(0) + num_tokens,
+                               tokens.begin(),
+                               token_id.begin(),
+                               is_nested,
+                               stream);
+
+    thrust::exclusive_scan(
+      rmm::exec_policy(stream), token_levels.begin(), token_levels.end(), token_levels.begin());
+
+    // Get parent of first child of struct/list begin.
+    auto const first_childs_parent_token_id2 =
```

Comment:
nit: Can we find a better name for `first_childs_parent_token_id2`? This is for nested types only, so maybe `nested_first_childs_parent_token_id`?

Reply:
renamed